### PR TITLE
Add support for sort filter #233

### PR DESCRIPTION
**Pull request by hoggetaylor**
I'm submitting this as is to get some feedback.

Some things I'm currently unhappy with:
- An attribute that does not map to an actual field on the object fails silently.
- Sorting a list of numbers that contains NaN or infinity will cause a panic. (See all of the unwraps on line 72 of array.rs)

### CODE REVIEW

File: src/builtins/filters/array.rs [hunk 1]
```diff
@@ -40,6 +41,53 @@ pub fn join(value: Value, args: HashMap<String, Value>) -> Result<Value> {
     Ok(to_value(&rendered.join(&sep))?)
 }
 
+/// Sorts the array in ascending order.
+/// Use the 'attribute' argument to define a field to sort by.
+/// Set the 'reverse' argument to sort in descending order.
```
Comment (Keats) [inline src/builtins/filters/array.rs]: There is already a `reverse` filter so there is no need for a `reverse` option

File: src/builtins/filters/array.rs [hunk 2]
```diff
@@ -40,6 +41,53 @@ pub fn join(value: Value, args: HashMap<String, Value>) -> Result<Value> {
...
+    use Value::*;
+    use self::Ordering::*;
+    match (a, b) {
+        (&Null, &Null) => Equal,
+        (&Bool(ref a), &Bool(ref b)) => a.cmp(b),
+        (&Number(ref a), &Number(ref b)) => a.as_f64().unwrap().partial_cmp(&b.as_f64().unwrap()).unwrap(),
+        (&String(ref a), &String(ref b)) => a.cmp(b),
+        (&Array(ref a), &Array(ref b)) => a.len().cmp(&b.len()),
+        (&Object(ref a), &Object(ref b)) => a.len().cmp(&b.len()),
+        (a, b) => type_of(a).cmp(&type_of(b))
```
Comment (Keats) [inline src/builtins/filters/array.rs]: I don't think we should sort things that cannot be objectively sorted.

**Comment by Keats** [comment]
cc @stephanbuys

**Comment by Keats** [comment]
@hoggetaylor are you still working on that?

**Comment by hoggetaylor** [comment]
Sorry I have been busy finishing up exams. I am still working in this. I hope to have something pushed up soon.

**Comment by hoggetaylor** [comment]
commit 6e29de6 resolves all panics in the previous implementation. Also giving an invalid attribute will result in an Err. 

The code clones a bit more than I would like but I was having some lifetime issues that I'm not quite experienced enough to resolve.

**Comment by Keats** [comment]
The sort looks good as well. I would maybe just put all the code unrelated to the sort filter in another file though.

@stephanbuys can you confirm that it is what you had in mind?

**Comment by Keats** [comment]
Thanks! I think it's just missing the docs in `docs/content/docs/templates.md` and it's good to merge

File: docs/content/docs/templates.md [hunk 3]
```diff
@@ -482,6 +482,43 @@ Returns the length of an array or a string, 0 if the value is not an array.
...
 Returns a reversed string or array.
 
+### sort
+Sorts an array into ascending order.
+
+The values in the array must be a sortable type:
+- numbers are sorted by their numerical value.
+- strings are sorted in alphabetical order.
+- arrays are sorted by their length.
+- bools are sorted as if false=0 and true=1
```
Comment (Keats) [inline docs/content/docs/templates.md]: now that i think of it, does it ever makes sense to sort an array of bool?

Comment (hoggetaylor) [reply docs/content/docs/templates.md]: It's probably a pretty rare use case. But I don't think it would hurt to have the ability to do it.

**Comment by Keats** [comment]
Thanks a lot!

**Pull request merged**